angular/src/app/admin/dam-repository/service-definitions/service-definition-form/service-definition-form.component.ts

```diff
@@ -0,0 +1,208 @@
+      }),
+    });
+  }
+
+  ngAfterViewInit(): void {
+    const roles = _get(this.serviceTemplate, 'dto.roles', {});
+    const interfaces = _get(this.serviceTemplate, 'dto.interfaces', {});
+    this.updateRolesForm(roles);
+    this.updateInterfacesForm(interfaces);
+    this.changeDetector.detectChanges();
+  }
+
+  addRole(roleId?: string, roles?): void {
+    if (!roles) {
+      const { name, dto } = new EntityModel( '' , ServiceRole.create());
+      this.roles.insert(0, this.createRoleForm(name, dto));
+    } else {
+      const {name, dto} = new EntityModel(roleId, _get(roles, roleId, null));
+      this.roles.insert(0, this.createRoleForm(name, dto));
+    }
+  }
+
+  removeRole(index: number): void {
+    this.roles.removeAt(index);
+  }
+
+  addInterface(interfaceType?, interfaces?) {
+    if (!interfaces) {
+      this.interfaces.insert(0, this.createInterfacesForm('', ''));
+    } else {
+      const {name, dto} = new EntityModel(interfaceType, _get(interfaces, interfaceType, null));
+      this.interfaces.insert(0, this.createInterfacesForm(name, dto));
+    }
+  }
+
+  removeInterface(index: number) {
+    this.interfaces.removeAt(index);
+  }
+
+  getModel(): EntityModel {
+    const {
+      id,
+      interfaces,
+      itemFormat,
+      roles,
+      targetAdapter,
+      ui,
+    } = this.form.value;
+    return new EntityModel(id, {
+      interfaces: this.formatInterfaces(interfaces),
+      itemFormat,
+      roles: this.formatRoles(roles),
+      targetAdapter,
+      ui,
+    });
+  }
+
+  targetAdapterChange() {
+    const selectedTargetAdapter: string = this.form.get('targetAdapter').value || '';
+    if (selectedTargetAdapter) {
+      this.itemFormats = this.targetAdapters[selectedTargetAdapter].itemFormats;
+      this.requirements = this.targetAdapters[selectedTargetAdapter].requirements;
+    }
+  }
+
+  getAllForms(): FormGroup[] {
+    return [this.form];
+  }
+
+  isValid(): boolean {
+    return this.form.valid;
+  }
+
+  private createRoleForm(name, dto): FormGroup {
+    return this.formBuilder.group({
+      name: [name, []],
+      targetRoles: [dto.targetRoles, []],
+      targetScopes: [dto.targetScopes, []],
+      damRoleCategories: [dto.damRoleCategories, []],
+      ui: this.formBuilder.group({
+        label: [ dto.ui.label, [Validators.required]],
+        description: [dto.ui.description, [Validators.required]],
+      }),
+    });
+  }
+
+  private createInterfacesForm(name, dto): FormGroup {
+    return this.formBuilder.group({
+      type: [name],
+      value: [dto],
+    });
+  }
+
+  private updateRolesForm(roles: any) {
+    if (!roles) {
+      return;
+    }
+    Object.keys(roles).forEach(roleId => this.addRole(roleId, roles));
+  }
+
+  private updateInterfacesForm(interfaces: any) {
+    if (!interfaces) {
+      return;
+    }
+    Object.keys(interfaces).forEach(interfaceType => this.addInterface(interfaceType, interfaces));
+  }
+
+  private formatInterfaces(interfaces) {
```

Comment:
instead of `format....` I would rename it to `getIterfacesModel`, `getRolesModel`....

Reply:
can you simplify this function?